clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	goerrors "github.com/go-errors/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+	"sync"
+	"time"
+)
+
+const (
+	apiAvailabilityName = "ApiAvailability"
+)
+
+func init() {
+	if err := measurement.Register(apiAvailabilityName, createApiAvailabilityMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", apiAvailabilityName, err)
+	}
+}
+
+func createApiAvailabilityMeasurement() measurement.Measurement {
+	return &apiAvailabilityMeasurement{}
+}
+
+type apiAvailabilityOutput struct {
+	AvailabilityPercentage     int `json:"availabilityPercentage"`
+	LongestUnavailableDuration int `json:"longestUnavailableDuration"`
+}
+
+type apiAvailabilityMetrics struct {
+	numProbes                  int
+	numSuccess                 int
+	numFailures                int
+	prevProbeFailure           bool
+	maxConsecutiveFailedProbes int
+	consecutiveFailedProbes    int
+}
+
+type apiAvailabilityMeasurement struct {
+	isRunning              bool
+	stopCh                 chan struct{}
+	hosts                  []string
+	apiAvailabilityMetrics *apiAvailabilityMetrics
+	wg                     sync.WaitGroup
+}
+
+func (a *apiAvailabilityMeasurement) apiServerAvailable(c clientset.Interface, SSHToMasterSupported bool, config *measurement.MeasurementConfig) bool {
+	for _, host := range a.hosts {
+		// Get the api-server measurement data over SSH.
+		// Start by checking that the provider allows us to do so.
+		if !SSHToMasterSupported {
```

Comment:
I don't think we should even consider SSH for this purpose.
Simply calling /healthz using client is what we want.

The problem is that using client you currently contact a random master. That said, maybe it's actually fine - if there is a single master down but load-balancer is not forwarding any requests to it, that sounds like a healthy cluster?

Thinking more about it, what we may consider is:
- healthiness at the cluster level
- healthiness of individual masters

Reply:
[That said, I'm not sure if in CL2 we support passing "cluster-level IP" and "master IPs". Maybe we do and the first is via kubeconfig?]

Reply:
> I don't think we should even consider SSH for this purpose.
> Simply calling /healthz using client is what we want.
> 
> The problem is that using client you currently contact a random master. That said, maybe it's actually fine - if there is a single master down but load-balancer is not forwarding any requests to it, that sounds like a healthy cluster?
> 
> Thinking more about it, what we may consider is:
> 
> * healthiness at the cluster level
> * healthiness of individual masters

As per my understanding, If a single master is down and LB isn't forwarding any requests to it, then it can't be qualified as a healthy cluster. 

I think with the current implementation, for non-ssh mode, what we are considering is cluster level health. So, if a request returns, we count it as a success. 
It's only in the ssh case, where we consider hosts but in this case, if any one of the hosts reports an error, it is counted as a failure. So, that's an individual master level probe. 

So, you are saying we report both cluster/individual level modes? If we add individual level, then we will also need to track which masters failed how many times?

Reply:
> As per my understanding, If a single master is down and LB isn't forwarding any requests to it, then it can't be qualified as a healthy cluster.

If there are multiple masters and one is down but not receiving any traffic, I claim that from user-perspective the cluster is healthy. And we care about user-perspective here.

> So, you are saying we report both cluster/individual level modes? If we add individual level, then we will also need to track which masters failed how many times?

yes - we need the same data for 1. cluster as a whole, 2. each master separately

Reply:
Understood. Do we want the health details for each master at every probe? If yes, then I will have to change the code - which I reused from the probes example that you had sent.
So, when SSH is not supported, then what we will have is only cluster level metrics.

It's only when ssh is supported that I can get individual level metrics. Here, currently from the re-used logic, if **any** of the master fails, the method returns a false for api server availability. Instead, I will have to ssh into each master, get the success/failure and report it back and update in a map where I will store metrics against each master. 

The only thing is, if we have too many masters then is there a chance of the probes overlapping with each other? Can that have a bearing on the metrics reported?

And lastly, is it a fair thing to say that if ssh isn't supported then all we get is cluster level metrics, and in the other case, both cluster level and individual master level metrics?

Reply:
I think we should start with something and iterate, rather than trying to get a perfect solution from the very beginning:
- I'm fine initially that only return all metrics for ssh-supported providers [and that should be hidden behind flag anyway]
- re overlapping probes - that also sounds fine at least initially 

A clear TODOs should be added though to emhpasize the problems we're aware of.

Reply:
hey @wojtek-t , I have made the changes. I couldn't push them because formatting changes are failing for a recently pushed commit to probes.go. I have rectified it here:
https://github.com/kubernetes/perf-tests/pull/1190

Once merged, I will push my changes.

Reply:
Thanks for merging. I pushed the latest changes.

Reply:
hi @wojtek-t , did you get a chance to look at the latest changes?